soc/arm/atmel_sam/same70/soc.c

```diff
@@ -226,16 +227,19 @@ static ALWAYS_INLINE void clock_init(void)
 
 void z_arm_platform_init(void)
 {
-	if (IS_ENABLED(CONFIG_CACHE_MANAGEMENT) && IS_ENABLED(CONFIG_ICACHE)) {
-		SCB_EnableICache();
-	} else {
-		SCB_DisableICache();
-	}
-	if (IS_ENABLED(CONFIG_CACHE_MANAGEMENT) && IS_ENABLED(CONFIG_DCACHE)) {
-		SCB_EnableDCache();
-	} else {
-		SCB_DisableDCache();
-	}
+	/*
+	 * DTCM is enabled by default at reset, therefore we have to disable
+	 * it first to get the caches into a state where then the
+	 * sys_cache*-functions can enable them, if requested by the
+	 * configuration.
+	 */
+	cache_data_disable();
+
+	/*
+	 * Enable the caches only if configured to do so.
+	 */
+	sys_cache_instr_enable();
+	sys_cache_data_enable();
```

Comment:
I think this still have issues:

The problem is that the API assumes that CONFIG_CACHE_MANAGEMENT is always enabled and that is not true. If the user say CONFIG_CACHE_MANAGEMENT=NO how the cache is disabled? Maybe it is easy to understand when a bootloader set I-cache TRUE and the system will never disabled because CONFIG_CACHE_MANAGEMENT=NO.

However I think it is safe to replace `SCB_*` by `arch_*` equivalent if that is acceptable.

BTW, I noted that Atmel is not clearing the cache and the system do not clear too for every enable/disable which is a bug.

Reply:
With this implementation both caches will be disabled if CONFIG_CACHE_MANAGEMENT=N, which is the intended result according to @carlocaione.
z_arm_platform_init will assume a clean start without a bootloader beforehand. For these cases, with a bootloader, I think INIT_ARCH_HW_AT_BOOT should be used, which in turn will trigger a call to INIT_ARCH_HW_AT_BOOT during startup. The implementation there looks faulty to me, as it is using the sys_cache*-functions. But this is a different topic, in my opinion.

Reply:
I agreed that this works only if the INIT_ARCH_HW_AT_BOOT is always enabled for the Atmel SAM after this change. Without that, the initialization can not ensure correct cache initialization if user is using SAM-BA bootloader (ROM bootloader) for instance or any other bootloader due to intrinsic initialization sequence.

So I'll ask you to add the following patch.

```diff
diff --git a/soc/arm/atmel_sam/same70/Kconfig.series b/soc/arm/atmel_sam/same70/Kconfig.series
index 66ceeb6ca3..4e7d6aa396 100644
--- a/soc/arm/atmel_sam/same70/Kconfig.series
+++ b/soc/arm/atmel_sam/same70/Kconfig.series
@@ -14,6 +14,7 @@ config SOC_SERIES_SAME70
        select CPU_HAS_ICACHE
        select CPU_HAS_DCACHE
        select SOC_FAMILY_SAM
+       select INIT_ARCH_HW_AT_BOOT
        select PLATFORM_SPECIFIC_INIT
        select ASF
        select HAS_SWO
diff --git a/soc/arm/atmel_sam/samv71/Kconfig.series b/soc/arm/atmel_sam/samv71/Kconfig.series
index 543def9d9a..cadee35acb 100644
--- a/soc/arm/atmel_sam/samv71/Kconfig.series
+++ b/soc/arm/atmel_sam/samv71/Kconfig.series
@@ -14,6 +14,7 @@ config SOC_SERIES_SAMV71
        select CPU_HAS_ICACHE
        select CPU_HAS_DCACHE
        select SOC_FAMILY_SAM
+       select INIT_ARCH_HW_AT_BOOT
        select PLATFORM_SPECIFIC_INIT
        select ASF
        select HAS_SWO
```

Reply:
done

Reply:
Hi @benediktibk ,

Thank you for your effort to address this.